std/range/package.d

```diff
@@ -2688,16 +2695,124 @@ pure @safe nothrow @nogc unittest
     struct NonForwardRange
     {
         enum empty = false;
-        int front() { return 42; }
+        @property int front() { return 42; }
         void popFront() {}
     }
 
     static assert(!isForwardRange!NonForwardRange);
 
     auto s = takeOne(NonForwardRange());
+    assert(s.length == 1);
+    assert(!s.empty);
     assert(s.front == 42);
+    assert(s.back == 42);
+    assert(s[0] == 42);
+
+    auto t = s[0 .. 0];
+    assert(t.empty);
+    assert(t.length == 0);
+
+    auto u = s[1 .. 1];
+    assert(u.empty);
+    assert(u.length == 0);
+
+    auto v = s[0 .. 1];
+    s.popFront();
+    assert(s.length == 0);
+    assert(s.empty);
+    assert(!v.empty);
+    assert(v.front == 42);
+    v.popBack();
+    assert(v.empty);
+    assert(v.length == 0);
 }
 
+pure @safe nothrow @nogc unittest
+{
+    struct NonSlicingForwardRange
+    {
+        enum empty = false;
+        @property int front() { return 42; }
+        void popFront() {}
+        @property auto save() { return this; }
+    }
+
+    static assert(isForwardRange!NonSlicingForwardRange);
+    static assert(!hasSlicing!NonSlicingForwardRange);
+
+    auto s = takeOne(NonSlicingForwardRange());
+    assert(s.length == 1);
+    assert(!s.empty);
+    assert(s.front == 42);
+    assert(s.back == 42);
+    assert(s[0] == 42);
+    auto t = s.save;
+    s.popFront();
+    assert(s.length == 0);
+    assert(s.empty);
+    assert(!t.empty);
+    assert(t.front == 42);
+    t.popBack();
+    assert(t.empty);
+    assert(t.length == 0);
+}
+
+// Test that asserts trigger correctly
+@system unittest
+{
+    import core.exception : AssertError;
+
+    struct NonForwardRange
+    {
+        enum empty = false;
+        @property int front() { return 42; }
+        void popFront() {}
+    }
+
+    auto s = takeOne(NonForwardRange());
+
+    bool thrown = false;
+    try s[1];
+    catch (AssertError) thrown = true;
+    assert(thrown);
+
+    thrown = false;
+    try s[0 .. 2];
+    catch (AssertError) thrown = true;
+    assert(thrown);
+
+    thrown = false;
+    size_t one = 1;     // Avoid style warnings triggered by literals
+    size_t zero = 0;
+    try s[one .. zero];
+    catch (AssertError) thrown = true;
+    assert(thrown);
+
+    s.popFront;
+    assert(s.empty);
+
+    thrown = false;
+    try s.front;
+    catch (AssertError) thrown = true;
+    assert(thrown);
+
+    thrown = false;
+    try s.back;
+    catch (AssertError) thrown = true;
+    assert(thrown);
+
+    thrown = false;
+    try s.popFront;
+    catch (AssertError) thrown = true;
+    assert(thrown);
+
+    thrown = false;
+    try s.popBack;
+    catch (AssertError) thrown = true;
+    assert(thrown);
```

Comment:
This happens fairly often. Why not move it in a function?
Sth. like e.g. works:

```d
import core.exception : AssertError;

void assertThrown(E)(lazy E exp)
{
    bool thrown;
    try
    {
        exp();
    }
    catch (AssertError)
    {
        thrown = true;
    }
    assert(thrown);
}

void main()
{
    assertThrown(assert(0));
    assertThrown(1 + 1);
}
```

https://run.dlang.io/is/rkYXbv

Reply:
Nice idea. I do think there is a small disadvantage of using a relatively uncommon language feature (lazy function parameter evaluation), but I agree that it would save some space in the code. After looking at a few of these, I'm starting to think more unit tests should be covering the assert cases.

Reply:
Turns out the it's not necessary to write a function. These tests can be written by passing `AssertError` as a template parameter to `assertThrown`. For example:
```
   assertThrown!AssertError(s.front);
```
These unit tests have been rewritten in this style.